mtenv/envs/registration.py

```diff
@@ -1,86 +1,244 @@
-
-        Args:
-            id (str): The official environment ID
-            entry_point (Optional[str]): The Python entrypoint of the
-                environment class (e.g. module.name:Class)
-            reward_threshold (Optional[int]): The reward threshold before
-                the task is considered solved
-            kwargs (dict): The kwargs to pass to the environment class
-            nondeterministic (bool): Whether this environment is
-                non-deterministic even after seeding
-            max_episode_steps (Optional[int]): The maximum number of steps
-                that an episode can consist of
-            test_kwargs (Optional[Dict[str, Any]], optional): Dictionary
-                to specify parameters for automated testing. Defaults to
-                None.
-
-        """
-        super().__init__(
-            id=id,
-            entry_point=entry_point,
-            reward_threshold=reward_threshold,
-            nondeterministic=nondeterministic,
-            max_episode_steps=max_episode_steps,
-            kwargs=kwargs,
-        )
-        self.test_kwargs = test_kwargs
+import copy
+import importlib
+import importlib.util
+from dataclasses import dataclass, field
+from typing import Any, Dict, Optional, Sequence, Union
+
+from gym import error, logger
+from gym.envs.registration import (  # noqa: F401
+    EnvSpec,
+    _check_spec_register,
+    _check_version_exists,
+    current_namespace,
+    find_highest_version,
+    get_env_id,
+    load,
+    parse_env_id,
+    registry,
+)
+from gym.wrappers import (  # type: ignore[attr-defined]
+    AutoResetWrapper,
+    HumanRendering,
+    OrderEnforcing,
+    StepAPICompatibility,
+    TimeLimit,
+)
+from gym.wrappers.env_checker import PassiveEnvChecker
+
+
+@dataclass
+class MultitaskEnvSpec(EnvSpec):
+
+    test_kwargs: Dict[str, Any] = field(default_factory=dict)
 
     def __repr__(self) -> str:
         return f"MultitaskEnvSpec({self.id})"
 
-    @property
-    def kwargs(self) -> Dict[str, Any]:
-        return self._kwargs  # type: ignore[no-any-return]
-
 
-class MultiEnvRegistry(EnvRegistry):  # type: ignore[misc]
-    def __init__(self) -> None:
-        super().__init__()
-
-    def register(self, id: str, **kwargs: Any) -> None:
-        if id in self.env_specs:
-            raise error.Error("Cannot re-register id: {}".format(id))
-        self.env_specs[id] = MultitaskEnvSpec(id, **kwargs)
+def register(id: str, **kwargs: Any) -> None:
+    """Register an environment with gym.
+
+    The `id` parameter corresponds to the name of the environment, with the syntax as follows:
+    `(namespace)/(env_name)-v(version)` where `namespace` is optional.
+
+    It takes arbitrary keyword arguments, which are passed to the `MultitaskEnvSpec` constructor.
+
+    Args:
+        id: The environment id
+        **kwargs: arbitrary keyword arguments which are passed to the environment constructor
+    """
+    global registry, current_namespace
```

Comment:
Why do you need these as `global`?

Edit: ah, I just realized it was a copy of the code from gym, with just a minor diff. Might be worth mentioning it in a comment. And same for `make()` below. This kind of stuff is always a bit annoying in Python, although sometimes hard to avoid. What's the main reason why we need to override `EnvSpec`?